### PR TITLE
Update Dockerfile to match llvm-premerge-checks.

### PR DESCRIPTION
**Pull request by pcf000**
Three purposes:
1. issue [#1476](https://github.com/ROCm/rocMLIR-internal/issues/1476)
2. fix spurious codecov failures
3. fix Stefan's problems with upstream merge

Update the first section of our Dockerfile with the latest from llvm-premerge-checks.

### CODE REVIEW

File: mlir/utils/jenkins/Dockerfile [hunk 1]
```diff
@@ -12,66 +12,79 @@ RUN apt-get clean all
 
 # --------------------- Section 1: ported from llvm-premerge-checks -----------------
 # Keep this section up-to-date with the upstream
-# https://github.com/google/llvm-premerge-checks/blob/master/containers/base-debian/Dockerfile
```
Comment (pcf000) [inline mlir/utils/jenkins/Dockerfile:15]: Most of these diffs are just changes in the upstream source.  I did remove installation of gosu, tini, autoconf, automake, and libtool that I don't think we use.

File: mlir/utils/jenkins/Dockerfile [hunk 2]
```diff
@@ -12,66 +12,79 @@ RUN apt-get clean all
...
+RUN wget --no-verbose -O /cmake.sh https://github.com/Kitware/CMake/releases/download/v3.23.3/cmake-3.23.3-linux-x86_64.sh; \
+    chmod +x /cmake.sh; \
+    mkdir -p /etc/cmake; \
+    /cmake.sh --prefix=/etc/cmake --skip-license; \
+    ln -s /etc/cmake/bin/cmake /usr/bin/cmake; \
+    cmake --version; \
+    rm /cmake.sh
+
+# LLVM must be installed after prerequisite packages.
+ENV LLVM_VERSION=17
```
Comment (pcf000) [inline mlir/utils/jenkins/Dockerfile:46]: Upstream has 16, but rocm 6.1 has 17 and so does Stefan.

File: mlir/utils/jenkins/Dockerfile [hunk 3]
```diff
@@ -12,66 +12,79 @@ RUN apt-get clean all
...
+    ln -s /usr/bin/clang++-${LLVM_VERSION} /usr/bin/clang++ && \
+    ln -s /usr/bin/clang-tidy-${LLVM_VERSION} /usr/bin/clang-tidy && \
+    ln -s /usr/bin/clang-tidy-diff-${LLVM_VERSION}.py /usr/bin/clang-tidy-diff && \
+    ln -s /usr/bin/clang-format-${LLVM_VERSION} /usr/bin/clang-format && \
+    ln -s /usr/bin/git-clang-format-${LLVM_VERSION} /usr/bin/git-clang-format && \
+    ln -s /usr/bin/clang-format-diff-${LLVM_VERSION} /usr/bin/clang-format-diff && \
+    ln -s /usr/bin/lld-${LLVM_VERSION} /usr/bin/lld && \
+    ln -s /usr/bin/lldb-${LLVM_VERSION} /usr/bin/lldb && \
+    ln -s /usr/bin/ld.lld-${LLVM_VERSION} /usr/bin/ld.lld && \
+    ln -s /usr/bin/llvm-profdata-${LLVM_VERSION} /usr/bin/llvm-profdata && \
```
Comment (pcf000) [inline mlir/utils/jenkins/Dockerfile:63]: Linked these three just for us.  First two are for codecov.

File: mlir/utils/jenkins/Dockerfile [hunk 4]
```diff
@@ -86,19 +99,15 @@ RUN groupadd -g 109 render
...
-RUN echo 'deb [signed-by=/usr/share/keyrings/kitware-archive-keyring.gpg] https://apt.kitware.com/ubuntu/ focal main' | \
-    tee /etc/apt/sources.list.d/kitware.list >/dev/null
-
 # Note instead of installing the latest, we should always manually bump cmake version when necessary.
 # This make sure that we don't accidentally use newer cmake features incompatible with our client.
-# As of 2022-08-17, the rocm/rock images has /opt/conda/bin/cmake at 3.22.1, so we'll match that.
-# Note that the mlir minimum is 3.18.
-RUN apt-get update && rm /usr/share/keyrings/kitware-archive-keyring.gpg && \
+# llvm-premerge-checks installs 3.23.3, see above.
+# As of 2022-08-17, the rocm/rock images has /opt/conda/bin/cmake at 3.22.1.
```
Comment (pcf000) [inline mlir/utils/jenkins/Dockerfile:105]: I don't know which image has /opt/conda so I can't verify that version.  external/llvm-project/mlir/CMakeLists.txt does have 3.20.0.

File: mlir/utils/jenkins/Dockerfile [hunk 5]
```diff
@@ -86,19 +99,15 @@ RUN groupadd -g 109 render
...
+# Note that the mlir minimum is 3.20.0.
+RUN apt-get update && \
   apt-get install -y --no-install-recommends \
   rocm-dev \
+  rocm-llvm-dev \
```
Comment (pcf000) [inline mlir/utils/jenkins/Dockerfile:110]: Rocm versions of llvm-profdata, llvm-cov, git-clang-format.  Basically the same as 17, but we had them in rocm-5.x so let's keep them.

File: mlir/utils/jenkins/Dockerfile [hunk 6]
```diff
@@ -135,5 +139,5 @@ RUN python3 -m venv /tuna-venv && . /tuna-venv/bin/activate && \
     python3 -m pip install -r tuna-requirements.txt --ignore-installed && \
     python3 -m pip install scipy pandas
 
-# Workaround for 6.0.0 rocprof not supporting navi3x:  add it to list.
+# Workaround for 6.1.1 rocprof not supporting navi3x:  add it to list.
```
Comment (pcf000) [inline mlir/utils/jenkins/Dockerfile:142]: Issue [#1496](https://github.com/ROCm/rocMLIR-internal/issues/1496) addresses this hack.  I have a prototype solution that needs a little work, and a way to measure bank conflicts.

**Comment by pcf000** [comment]
I ran a nightly equivalent, selected tuning, and premerge checks, so I have some confidence it won't break things, but of course I'll deal with it if it does.

**Comment by stefankoncarevic** [comment]
I've reviewed the Dockerfile changes, and they look good. The updates align with the latest llvm-premerge-checks configurations and ensure the use of up-to-date tools and dependencies. 
I approve this PR. 
Thank you Paul.

**Review by stefankoncarevic — APPROVED** [review]
I've reviewed the Dockerfile changes, and they look good. The updates align with the latest llvm-premerge-checks configurations and ensure the use of up-to-date tools and dependencies.
I approve this PR.
Thank you Paul.

File: mlir/utils/jenkins/Dockerfile [hunk 7]
```diff
@@ -12,66 +12,79 @@ RUN apt-get clean all
...
+        # for libc++ tests that use the timezone database of the chrono header
+        tzdata \
+        # for llvm installation script
+        sudo \
+        # build scripts
+        nodejs ccache \
+        # shell users
+        less vim
+
+# Install cmake 3.23+ from source.
```
Comment (krzysz00) [inline mlir/utils/jenkins/Dockerfile:36]: Why? I'd install the package from kitware

File: mlir/utils/jenkins/Dockerfile (hunk 1 above)
Comment (krzysz00) [reply mlir/utils/jenkins/Dockerfile:15]: Also, should we move to UBuntu 22.04 here?

File: mlir/utils/jenkins/Dockerfile (hunk 7 above)
Comment (pcf000) [reply mlir/utils/jenkins/Dockerfile:36]: Ask google;  this is copied from llvm-premerge-checks's Dockerfile.

**Comment by pcf000** [comment]
I think 22.04 is beyond the scope here, but we could if there's demand.  I was going to say we "just" moved to 20.04, because it feels recent, but git-blame tells me it was November 2022.

**Review by jerryyin — APPROVED** [review]
No objections as long as CI can continue to pass. Because of the scale of change, I'm wondering if you got a chance to run some of the CI steps manually and make sure things still work?

**Comment by pcf000** [comment]
Pull-request CI took ForEvEr to get a navi3x, so it's still not done yet.  While it was churning, I experimented with advancing to ubuntu 22.04.  Not quite trivial -- took me a while to understand the apt "pin" concept -- but it does pass the same things.  I'll hold off on 22 for at least a few days, or longer if we'd rather stick with 20.

**Pull request merged**